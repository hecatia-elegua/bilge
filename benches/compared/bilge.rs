use bilge::{Bitsized, Number, FromBits, bitsize, u4, u12, u20, u5, u2, DebugBits};

#[inline(never)]
pub fn bilge(input: (u32, u32, u64, u16)) {
    let mut lpi = GicRedistributorLpi {
        control: RedistributorControl::from(input.0),
        implementer_identification: RedistributorImplementerIdentification::from(input.1),
        redistributor_type: RedistributorType::from(input.2),
    };
    assert_eq!(lpi.control.value, input.0);
    assert_eq!(lpi.implementer_identification.value, input.1);
    assert_eq!(lpi.redistributor_type.value, input.2);
    
    assert!(lpi.control.clear_enable_supported());
    assert_eq!(lpi.implementer_identification.implementer_jep106(), u12::new(2054));
    lpi.implementer_identification.set_implementer_jep106(u12::new(input.3));
    assert_eq!(lpi.implementer_identification.implementer_jep106(), u12::new(input.3));
    assert_eq!(lpi.redistributor_type.processor_number(), 63872);
}

#[derive(Debug)]
pub struct GicRedistributorLpi {
    control: RedistributorControl,
    implementer_identification: RedistributorImplementerIdentification,
    redistributor_type: RedistributorType,
}

#[bitsize(32)]
#[derive(DebugBits, FromBits)]
struct RedistributorControl {
    //ro
    upstream_write_pending: bool,
    //zero
    reserved: u4,
    //or reserved
        disable_processor_selection_for_group_1_secure_interrupts: bool,
        disable_processor_selection_for_group_1_non_secure_interrupts: bool,
        disable_processor_selection_for_group_0_interrupts: bool,
    //zero
    reserved: u20,
    register_write_pending: bool,
    lpi_invalidate_registers_supported: bool,
    clear_enable_supported: bool,
    enable_lpis: bool,
}

#[bitsize(32)]
#[derive(DebugBits, FromBits)]
struct RedistributorImplementerIdentification {
    //ro
    product_id: u8,
    //zero
    reserved: u4,
    //ro
    variant: u4,
    //ro
    revision: u4,
    implementer_jep106: u12,
}

#[bitsize(64)]
#[derive(DebugBits, FromBits)]
struct RedistributorType {
    affinity_value: u32,
    ppi_num: u5,
    virtual_sgi_supported: bool,
    common_lpi_affinity: u2, //TODO: enums everywhere
    processor_number: u16,
    resident_vpe_id: bool,
    mpam_supported: bool,
    control_cpgs_supported: bool,
    is_last: bool,
    direct_lpi_supported: bool,
    dirty: bool,
    virtual_lpi_supported: bool,
    physical_lpi_supported: bool,
}
