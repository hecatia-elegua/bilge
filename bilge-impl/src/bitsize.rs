use proc_macro2::{TokenStream, Ident};
use proc_macro_error::{abort_call_site, abort};
use quote::{quote, ToTokens};
use syn::{Item, ItemStruct, ItemEnum, Type, Attribute, Fields, Meta, parse_quote, spanned::Spanned};

use crate::shared::{self, BitSize, unreachable};

/// As `#[repr(u128)]` is unstable and currently no real usecase for higher sizes exists, the maximum is u64.
const MAX_ENUM_BIT_SIZE: BitSize = 64;

/// Since we want to be maximally interoperable, we need to handle attributes in a special way.
/// We use `#[bitsize]` as a sort of scope for all attributes below it and
/// the whole family of `-Bits` macros only works when used in that scope.
/// 
/// Let's visualize why this is the case, starting with some user-code:
/// ```ignore
/// #[bitsize(6)]
/// #[derive(Clone, Copy, PartialEq, DebugBits, FromBits)]
/// struct Example {
///     field1: u2,
///     field2: u4,
/// }
/// ```
/// First, the attributes get sorted, depending on their name.
/// Every attribute in need of field information gets resolved first,
/// in this case `DebugBits` and `FromBits`.
/// 
/// Now, after resolving all `before_compression` attributes, the halfway-resolved
/// code looks like this:
/// ```ignore
/// #[bilge::bitsize_internal(6)]
/// #[derive(Clone, Copy, PartialEq)]
/// struct Example {
///     field1: u2,
///     field2: u4,
/// }
/// ```
/// This `#[bitsize_internal]` attribute is the one actually doing the compression and generating
/// all the getters, setters and a constructor.
/// 
/// Finally, the struct ends up like this (excluding the generated impl blocks):
/// ```ignore
/// struct Example {
///     value: u6,
/// }
/// ```
struct SplitAttributes {
    before_compression: Vec<Attribute>,
    after_compression: Vec<Attribute>,
}

/// Intermediate Representation, just for bundling these together
struct ItemIr {
    name: Ident,
    /// needed in from_bits and try_from_bits
    filled_check: TokenStream,
    /// generated item (and size check)
    expanded: TokenStream,
}

pub(super) fn bitsize(args: TokenStream, item: TokenStream) -> TokenStream {
    let (item, declared_bitsize) = parse(item, args);
    let attrs = split_attributes(&item);
    let ir = match item {
        Item::Struct(mut item) => {
            modify_special_field_names(&mut item.fields);
            let name = item.ident.clone();
            let filled_check = analyze_struct(&item.fields);
            let expanded = generate_struct(&item, declared_bitsize);
            ItemIr { name, filled_check, expanded }
        }
        Item::Enum(item) => {
            let name = item.ident.clone();
            let filled_check = analyze_enum(declared_bitsize, item.variants.len());
            let expanded = generate_enum(&item);
            ItemIr { name, filled_check, expanded }
        }
        _ => unreachable(()),
    };
    generate_common(ir, attrs, declared_bitsize)
}

fn parse(item: TokenStream, args: TokenStream) -> (Item, BitSize) {
    let item = syn::parse2(item).unwrap_or_else(unreachable);

    if args.is_empty() {
        abort_call_site!("missing attribute value"; help = "you need to define the size like this: `#[bitsize(32)]`")
    }
    
    let (declared_bitsize, _arb_int) = shared::bitsize_and_arbitrary_int_from(args);
    if declared_bitsize > shared::MAX_STRUCT_BIT_SIZE {
        abort_call_site!("attribute is not a valid number"; help = "currently, numbers from 1 to {} are allowed", shared::MAX_STRUCT_BIT_SIZE)
    }
    (item, declared_bitsize)
}

/// Split item attributes into those applied before bitfield-compression and those applied after.
/// Also, abort on any invalid configuration.
fn split_attributes(item: &Item) -> SplitAttributes {
    match item {
        //enums don't need special handling
        Item::Enum(item) => SplitAttributes { 
            before_compression: item.attrs.clone(),
            after_compression: vec![]
        },
        Item::Struct(item) => {
            let mut from_bytes = None;
            let mut has_frombits = false;
            let mut before_compression = vec![];
            let mut after_compression = vec![];
            for attr in &item.attrs {
                if attr.to_token_stream().to_string().contains("bitsize_internal") {
                    abort!(attr, "remove bitsize_internal"; help = "attribute bitsize_internal can only be applied internally by the bitsize macros")
                }
                match &attr.meta {
                    Meta::List(list) => {
                        if !list.path.is_ident("derive") {
                            // It is most probable that basic attr macros work if we put them on after compression
                            after_compression.push(attr.clone());
                            continue;
                        }
                        attr.parse_nested_meta(|meta| {
                            let derive_path = meta.path;
                            if derive_path.is_ident("Debug") {
                                abort!(derive_path, "use DebugBits for structs")
                            }
                            let derive = parse_quote!(#[derive(#derive_path)]);

                            let derive_path_str = derive_path.get_ident().unwrap_or_else(|| {
                                // We could just use the last path segment or use `derive_str.contains()` but that sounds breakable.
                                // Handling this for real might be easy, I just don't know how right now.
                                abort!(derive_path, "we currently only support simple derives, without paths.");
                            }).to_string();
                            match derive_path_str.as_str() {
                                "FromBytes" => {
                                    from_bytes = Some(derive_path);
                                    after_compression.push(derive);
                                }
                                "FromBits" => {
                                    has_frombits = true;
                                    before_compression.push(derive)
                                }
                                "DebugBits" | "TryFromBits" => before_compression.push(derive),
                                // It is most probable that basic derive macros work if we put them on after compression
                                _ => after_compression.push(derive)
                            }
                            Ok(())
                        }).unwrap_or_else(unreachable)
                    }
                    // I don't know with which attrs I can hit Path and NameValue,
                    // so let's just put them on after compression.
                    _ => after_compression.push(attr.clone()),
                }
            }
            if let Some(from_bytes) = from_bytes {
                if !has_frombits {
                    abort!(from_bytes, "a bitfield struct with zerocopy::FromBytes also needs to have FromBits")
                }
            }
            SplitAttributes { before_compression, after_compression }
        },
        _ => abort_call_site!("item is not a struct or enum"; help = "`#[bitsize]` can only be used on structs and enums")
    }
}

fn generate_filled_check_for(ty: &Type) -> TokenStream {
    if shared::is_always_filled(ty) {
        return quote!(true);
    }
    use Type::*;
    match ty {
        // These don't work with structs or aren't useful in bitfields.
        BareFn(_) | Group(_) | ImplTrait(_) | Infer(_) | Macro(_) | Never(_) |
        // We could provide some info on error as to why Ptr/Reference won't work due to safety.
        Ptr(_) | Reference(_) |
        // The bitsize must be known at compile time.
        Slice(_) |
        // Something to investigate, but doesn't seem useful/usable here either.
        TraitObject(_) |
        // I have no idea where this is used.
        Verbatim(_) | Paren(_) => abort!(ty, "This field type is not supported"),
        Tuple(tuple) => {
            tuple.elems.iter().map(generate_filled_check_for)
                .reduce(|acc, next| quote!((#acc && #next)))
                // `field: (),` will be handled like this:
                .unwrap_or_else(|| quote!(true))
        },
        Array(array) => {
            generate_filled_check_for(&array.elem)
        },
        Path(_) => {
            quote!(#ty::FILLED)
        },
        _ => abort!(ty, "This field type is currently not supported"),
    }
}

/// Allows you to give multiple fields the name `reserved` or `padding`
/// by numbering them for you.
fn modify_special_field_names(fields: &mut Fields) {
    // We could have just counted up, i.e. `reserved_0`, but people might interpret this as "reserved to zero".
    // Using some other, more useful unique info as postfix would be nice.
    // Also, it might be useful to generate no getters or setters for these fields and skipping some calc.
    let mut reserved_count = 0;
    let mut padding_count = 0;
    fields.iter_mut().for_each(|f| {
        if let Some(name) = &mut f.ident {
            if name == "reserved" || name == "_reserved" {
                reserved_count += 1;
                let span = name.span();
                let name = format!("reserved_{}", "i".repeat(reserved_count));
                f.ident = Some(Ident::new(&name, span))
            } else if name == "padding" || name == "_padding" {
                padding_count += 1;
                let span = name.span();
                let name = format!("padding_{}", "i".repeat(padding_count));
                f.ident = Some(Ident::new(&name, span))
            }
        }
    });
}

fn analyze_struct(fields: &Fields) -> TokenStream {
    // NEVER move this, since we validate all nested field types here as well.
    // If we do want to move this, add a new function just for validation.
    fields.iter()
        .map(|field| generate_filled_check_for(&field.ty))
        .reduce(|acc, next| quote!(#acc && #next))
        //when we only have uints or nothing as fields, return true
        .unwrap_or_else(|| quote!(true))
}

fn analyze_enum(bitsize: BitSize, variants_count: usize) -> TokenStream {
    if bitsize > MAX_ENUM_BIT_SIZE {
        abort_call_site!("enum bitsize is limited to 64")
    }
    let max_variants_count = 1u128 << bitsize;
    let enum_is_filled = variants_count as u128 == max_variants_count;
    quote!(#enum_is_filled)
}

fn generate_struct(item: &ItemStruct, declared_bitsize: u8) -> TokenStream {
    let ItemStruct { vis, ident, fields, .. } = item;
    let declared_bitsize = declared_bitsize as usize;

    let computed_bitsize = fields.iter().fold(quote!(0), |acc, next| {
        let field_size = shared::generate_type_bitsize(&next.ty);
        quote!(#acc + #field_size)
    });

    // we could remove this if the whole struct gets passed
    let is_tuple_struct = fields.iter().any(|field| field.ident.is_none());
    let fields_def = if is_tuple_struct {
        let fields = fields.iter();
        quote! {
            ( #(#fields,)* );
        }
    } else {
        let fields = fields.iter();
        quote! {
            { #(#fields,)* }
        }
    };

    quote! {
        #vis struct #ident #fields_def

        // constness: when we get const blocks evaluated at compile time, add a const computed_bitsize
        const _: () = assert!(
            (#computed_bitsize) == (#declared_bitsize),
            concat!("struct size and declared bit size differ: ",
            // stringify!(#computed_bitsize),
            " != ",
            stringify!(#declared_bitsize))
        );
    }
}

// attributes are handled in `generate_common`
fn generate_enum(item: &ItemEnum) -> TokenStream {
    let ItemEnum { vis, ident, variants, .. } = item;
    quote! {
        #vis enum #ident {
            #variants
        }
    }
}

/// we have _one_ generate_common function, which holds everything that struct and enum have _in common_.
/// Everything else has its own generate_ functions.
fn generate_common(ir: ItemIr, attrs: SplitAttributes, declared_bitsize: u8) -> TokenStream {
    let ItemIr { name: item_type, filled_check, expanded } = ir;
    let SplitAttributes { before_compression, after_compression } = attrs;

    let bitsize_internal_attr =  quote! {#[bilge::bitsize_internal(#declared_bitsize)]};

    quote! {
        #(#before_compression)*
        #bitsize_internal_attr
        #(#after_compression)*
        #expanded

        impl #item_type {
            const FILLED: bool = #filled_check;
        }
    }
}
