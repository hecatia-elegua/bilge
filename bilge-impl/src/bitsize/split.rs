use proc_macro_error::{abort_call_site, abort};
use quote::ToTokens;
use syn::{meta::ParseNestedMeta, Path, Item, Attribute, Meta, parse_quote};
use crate::shared::{unreachable, util::PathExt};

/// Since we want to be maximally interoperable, we need to handle attributes in a special way.
/// We use `#[bitsize]` as a sort of scope for all attributes below it and
/// the whole family of `-Bits` macros only works when used in that scope.
/// 
/// Let's visualize why this is the case, starting with some user-code:
/// ```ignore
/// #[bitsize(6)]
/// #[derive(Clone, Copy, PartialEq, DebugBits, FromBits)]
/// struct Example {
///     field1: u2,
///     field2: u4,
/// }
/// ```
/// First, the attributes get sorted, depending on their name.
/// Every attribute in need of field information gets resolved first,
/// in this case `DebugBits` and `FromBits`.
/// 
/// Now, after resolving all `before_compression` attributes, the halfway-resolved
/// code looks like this:
/// ```ignore
/// #[bilge::bitsize_internal(6)]
/// #[derive(Clone, Copy, PartialEq)]
/// struct Example {
///     field1: u2,
///     field2: u4,
/// }
/// ```
/// This `#[bitsize_internal]` attribute is the one actually doing the compression and generating
/// all the getters, setters and a constructor.
/// 
/// Finally, the struct ends up like this (excluding the generated impl blocks):
/// ```ignore
/// struct Example {
///     value: u6,
/// }
/// ```
pub struct SplitAttributes {
    pub before_compression: Vec<Attribute>,
    pub after_compression: Vec<Attribute>,
}

impl SplitAttributes {
    /// Split item attributes into those applied before bitfield-compression and those applied after.
    /// Also, abort on any invalid configuration.
    /// 
    /// Any derives with suffix `Bits` will be able to access field information.
    /// This way, users of `bilge` can define their own derives working on the uncompressed bitfield.
    pub fn from_item(item: &Item) -> SplitAttributes {
        let attrs = match item {
            Item::Enum(item) => &item.attrs,
            Item::Struct(item) => &item.attrs,
            _ => abort_call_site!("item is not a struct or enum"; help = "`#[bitsize]` can only be used on structs and enums"),
        };

        let parsed = attrs.iter().map(parse_attribute);
        
        let is_struct = matches!(item, Item::Struct(..));
        
        let mut from_bytes = None;
        let mut has_frombits = false;

        let mut before_compression = vec![];
        let mut after_compression = vec![];

        for parsed_attr in parsed {
            match parsed_attr {
                ParsedAttribute::DeriveList(derives) => {
                    for mut derive in derives {
                        if derive.matches(&["zerocopy", "FromBytes"]) {
                            from_bytes = Some(derive.clone());
                        } else if derive.matches(&["bilge", "FromBits"]) {
                            has_frombits = true;
                        } else if derive.matches_core_or_std(&["fmt", "Debug"]) && is_struct {
                            abort!(derive.0, "use derive(DebugBits) for structs")
                        } else if derive.matches_core_or_std(&["default", "Default"]) && is_struct {
                            // emit_warning!(derive.0, "use derive(DefaultBits) for structs")
                            derive.0 = syn::parse_quote!(::bilge::DefaultBits);
                        }
                    
                        if derive.is_custom_bitfield_derive() {
                            before_compression.push(derive.into_attribute());
                        } else {
                            // It is most probable that basic derive macros work if we put them on after compression
                            after_compression.push(derive.into_attribute());
                        }
                    }
                },

                ParsedAttribute::BitsizeInternal(attr) => {
                    abort!(attr, "remove bitsize_internal"; help = "attribute bitsize_internal can only be applied internally by the bitsize macros")
                },

                ParsedAttribute::Other(attr) => {
                    // I don't know with which attrs I can hit Path and NameValue,
                    // so let's just put them on after compression.
                    after_compression.push(attr.to_owned())
                },
            };
        }

        if let Some(from_bytes) = from_bytes {
            if !has_frombits {
                abort!(from_bytes.0, "a bitfield with zerocopy::FromBytes also needs to have FromBits")
            }
        }

        // currently, enums don't need special handling - so just put all attributes before compression
        if !is_struct {
            before_compression.append(&mut after_compression)
        }
        
        SplitAttributes { before_compression, after_compression }    
    }
}

fn parse_attribute(attribute: &Attribute) -> ParsedAttribute {
    match &attribute.meta {
        Meta::List(list) if list.path.is_ident("derive") => {
            let mut derives = Vec::new();
            let add_derive = |meta: ParseNestedMeta| {
                let derive = Derive(meta.path);
                derives.push(derive);

                Ok(())
            };

            list.parse_nested_meta(add_derive).unwrap_or_else(|e| abort!(list.tokens, "failed to parse derive: {}", e));

            ParsedAttribute::DeriveList(derives)
        }

        meta if contains_anywhere(meta, "bitsize_internal") => ParsedAttribute::BitsizeInternal(attribute),

        _ => ParsedAttribute::Other(attribute),
    }
}

/// a crude approximation of things we currently consider in item attributes
enum ParsedAttribute<'attr> {
    DeriveList(Vec<Derive>),
    BitsizeInternal(&'attr Attribute),
    Other(&'attr Attribute),
}

/// the path of a single derive attribute, parsed from a list which may have contained several
#[derive(Clone)]
struct Derive(Path);

impl Derive {
    /// a new `#[derive]` attribute containing only this derive
    fn into_attribute(self) -> Attribute {
        let path = self.0;
        parse_quote! { #[derive(#path)] }
    }

    /// by `bilge` convention, any derive satisfying this condition is able
    /// to access bitfield structure information pre-compression, 
    /// allowing for user derives
    fn is_custom_bitfield_derive(&self) -> bool {
        let last_segment = self.0.segments.last().unwrap_or_else(|| unreachable(()));

        last_segment.ident.to_string().ends_with("Bits")
    }
}

impl PathExt for Derive {
    fn matches(&self, str_segments: &[&str]) -> bool {
        self.0.matches(str_segments)
    }
}

/// slightly hacky. attempts to recognize cases where an ident is deeply-nested in the meta.
fn contains_anywhere(meta: &Meta, ident: &str) -> bool {
    meta.to_token_stream().to_string().contains(ident)
}