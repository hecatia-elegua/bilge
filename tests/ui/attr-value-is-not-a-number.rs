use bilge::bitsize;

#[bitsize(...)]
enum Test {}

fn main() {}
