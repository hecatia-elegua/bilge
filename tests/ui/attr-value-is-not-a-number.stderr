error: attribute value is not a number
 --> tests/ui/attr-value-is-not-a-number.rs:3:11
  |
3 | #[bitsize(...)]
  |           ^^^
  |
  = help: you need to define the size like this: `#[bitsize(32)]`
